Address PROTOTYPE comments around ParenthesizedOperation and not eras…

333fred opened the pull request:
…ing UnaryOperations uncecessarily.

@dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2096,8 +2096,10 @@ private IOperation VisitConditionalExpression(IOperation condition, bool sense,
 
-            // PROTOTYPE(dataflow): Unwrap parenthesized and look for other places where it should happen
-            // PROTOTYPE(dataflow): Do not erase UnaryOperatorKind.Not if ProduceIsSense below will have to add it back.
+            while (condition.Kind == OperationKind.Parenthesized)
```
>while (condition.Kind == OperationKind.Parenthesized) [](start = 12, length = 53)

It feels like this should be done inside the loop too, after each time we dig through a logical Not. Consider changing the loop below to:

while(true)
{
    Dig through parentesized
    Break if not logical not
    Dig through logical Not
}

AlekseyTs replied:
Please add relevant test.

---
In reply to: [193943036](https://github.com/dotnet/roslyn/pull/27611#discussion_r193943036) [](ancestors = 193943036)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2106,6 +2108,20 @@ private IOperation VisitConditionalExpression(IOperation condition, bool sense,
+
+                // ProduceIsSense will add a UnaryOperator if sense is false, so don't bother erasing
+                // this one if it would need to add it back in
+                if (unOp.Operand.Kind != OperationKind.UnaryOperator &&
```
>if (unOp.Operand.Kind != OperationKind.UnaryOperator && [](start = 16, length = 55)

Consider if the following approach is simpler and involves less logic duplication:
- Remember the last Not that we about to dig through here.
- pass it to the ProduceIsSense
- Change ProduceIsSense to simply update the operand of that logical Not if we have it and need to invert the value.

BTW, it looks like ProduceIsSense can be inlined into this function because there are no more consumers.

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IBinaryOperatorExpression.vb:
```diff
@@ -7058,5 +7058,70 @@ Block[B5] - Exit
+            Dim source = <![CDATA[
+Public Class Class1
+    Sub M(a As Boolean, b As Boolean, c As Boolean)'BIND:"Sub M(a As Boolean, b As Boolean, c As Boolean)"
+        a = b AndAlso (Not c)
```
>a = b AndAlso (Not c) [](start = 8, length = 21)

This scenario looks sufficiently good to demonstrate the reuse of existing Not operator, but I think it doesn't quite show the impact from skipping parenthesized.
Try adding tests for ```a = b OrElse (Not c andalso Not d)``` and ```a = b OrElse Not (c orElse d)```. The should produce graphs equivalent to ```a = b OrElse Not c andalso Not d```, modulo the syntax associated with nodes.

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2065,7 +2065,7 @@ private IOperation VisitUserDefinedBinaryConditionalOperator(IBinaryOperation bi
             int captureId = captureIdForResult ?? _captureIdDispenser.GetNextId();
 
-            ref BasicBlockBuilder lazyFallThrough = ref stopSense ? ref fallToTrueOpt : ref fallToFalseOpt;
+            ref BasicBlockBuilder lazyFallThrough = ref stopValue ? ref fallToTrueOpt : ref fallToFalseOpt;
```
@AlekseyTs this was the bug. Rather than stopSense (which is true in the case of something like `!(b || c)`), we need to be looking at the actual desired value at the end.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [395639442](https://github.com/dotnet/roslyn/pull/27611#issuecomment-395639442) [](ancestors = 395639442)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2131,27 +2148,6 @@ private static bool IsBooleanConditionalOperator(IBinaryOperation binOp)
-                                                   semanticModel: null,
-                                                   condition.Syntax,
-                                                   condition.Type,
-                                                   constantValue: default, // revert constant value if we have one.
```
>// revert constant value if we have one. [](start = 75, length = 40)

Please open an issue for this comment.

AlekseyTs commented:
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

AlekseyTs commented:
@cston, @dotnet/roslyn-compiler Please review, need second review.

AlekseyTs commented:
Test windows_debug_vs-integration_prtest please
